File: DSA/Gautam_Shyamnani_590013078/Day26Question3.cpp
#include <iostream>
using namespace std;


struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};


Node* sortList(Node* head) {
    if (!head || !head->next) return head;

    Node *zeroHead = new Node(-1), *oneHead = new Node(-1), *twoHead = new Node(-1);
    Node *zero = zeroHead, *one = oneHead, *two = twoHead;

    Node* current = head;
    while (current) {
        if (current->data == 0) {
            zero->next = current;
            zero = zero->next;
        } else if (current->data == 1) {
            one->next = current;
            one = one->next;
        } else {
            two->next = current;
            two = two->next;
        }
        current = current->next;
    }


    zero->next = oneHead->next ? oneHead->next : twoHead->next;
    one->next = twoHead->next;
    two->next = nullptr;

    
    Node* sortedHead = zeroHead->next;

    delete zeroHead;
    delete oneHead;
    delete twoHead;

    return sortedHead;
}

void printList(Node* head) {
    while (head) {
        cout << head->data << " -> ";
        head = head->next;
    }
    cout << "nullptr\n";
}


Node* createList(int arr[], int n) {
    if (n == 0) return nullptr;
    Node* head = new Node(arr[0]);
    Node* tail = head;
    for (int i = 1; i < n; ++i) {
        tail->next = new Node(arr[i]);
        tail = tail->next;
    }
    return head;
}


int main() {
    int arr[] = {1, 2, 2, 1, 2, 0, 2, 2};
    int n = sizeof(arr) / sizeof(arr[0]);

    Node* head = createList(arr, n);

    cout << "Original List:\n";
    printList(head);

    head = sortList(head);

    cout << "\nSorted List:\n";
    printList(head);

    return 0;
}

File: DSA/N_PARTH_PANDEY_590012761/Day26Question3.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def sortList(head):
    if not head or not head.next:
        return head
    
    zero_head = ListNode()
    one_head = ListNode()
    two_head = ListNode()
    
    zero = zero_head
    one = one_head
    two = two_head
    
    current = head
    while current:
        if current.val == 0:
            zero.next = current
            zero = zero.next
        elif current.val == 1:
            one.next = current
            one = one.next
        else:
            two.next = current
            two = two.next
        current = current.next
    
    zero.next = one_head.next if one_head.next else two_head.next
    one.next = two_head.next
    two.next = None
    
    return zero_head.next

def print_list(head):
    current = head
    while current:
        print(current.val, end=" → ")
        current = current.next
    print("None")

# Create the linked list: 1 → 2 → 2 → 1 → 2 → 0 → 2 → 2
head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(2)
head.next.next.next = ListNode(1)
head.next.next.next.next = ListNode(2)
head.next.next.next.next.next = ListNode(0)
head.next.next.next.next.next.next = ListNode(2)
head.next.next.next.next.next.next.next = ListNode(2)

print("Input:")
print_list(head)

sorted_head = sortList(head)

print("Output:")
print_list(sorted_head)